p2p/service.py

```diff
@@ -0,0 +1,53 @@
+import asyncio
+import logging  # noqa: F401
+
+from p2p.cancel_token import CancelToken
+from p2p.exceptions import OperationCancelled
+
+
+class BaseService:
+    # Subclasses should define their own logger.
+    logger = None  # type: logging.Logger
+    # Number of seconds stop() will wait for run() to finish.
+    _wait_until_finished_timeout = 2
+
+    def __init__(self, token: CancelToken) -> None:
+        self._finished = asyncio.Event()
+        self.cancel_token = token
+
+    async def run(self) -> None:
+        """Awaits for the subclass' _run() coroutine and sets the _finished event."""
+        try:
+            await self._run()
+        except OperationCancelled as e:
+            self.logger.info("%s finished: %s", self, e)
+        except Exception:
+            self.logger.exception("Unexpected error in %s, exiting", self)
+        finally:
+            self._finished.set()
+
+    async def stop(self) -> None:
+        """Triggers the CancelToken and runs the subclass' _stop() coroutine.
+
+        Also waits for up to _wait_until_finished_timeout seconds for the run() coroutine to
+        finish.
+        """
+        self.cancel_token.trigger()
+        await self._stop()
+        try:
+            await asyncio.wait_for(self._finished.wait(), timeout=self._wait_until_finished_timeout)
+        except asyncio.futures.TimeoutError:
+            self.logger.info("Timed out waiting for %s to finish, exiting anyway", self)
+
+    #
+    # These methods should be implemented by subclasses
+    #
+    async def _run(self) -> None:
+        """Runs the service's loop.
+
+        Should return when the CancelToken is triggered.
```

Comment:
I interpret this comment as `_run` should catch `OperationCancelled` exceptions and return cleanly in those cases. But would this be necessary considering that `run` catches them itself and logs a nice message as well?

Reply:
The idea is that `_run()` can either return or let `OperationCancelled` bubble up, so I'll amend the docstring to make that clear